lib/tapioca/dsl/compilers/url_helpers.rb

```diff
@@ -136,15 +145,16 @@ def gather_non_discoverable_includers
 
           sig { params(mod: Module, helper: Module).returns(T::Boolean) }
           private def includes_helper?(mod, helper)
-            superclass_ancestors = []
+            return ancestors_of(mod).include?(helper) unless Class === mod
+
+            superclass = superclass_of(mod)
 
-            if Class === mod
-              superclass = superclass_of(mod)
-              superclass_ancestors = ancestors_of(superclass) if superclass
+            ancestors_of(mod).each do |a|
+              return true if helper.equal?(a)
+              return false if superclass.equal?(a)
```

Comment:
This hand-roll loop was the fastest implementation I could think of.

I also tried `ancestors_of(mod).take_while { |a| !superclass.equal?(a) }.include?(helper)`, but the extra array allocation/copying made it slower.

Reply:
I guess the assumption here is that we can stop the search when we hit the `superclass` in the ancestor chain, but I am not sure if that is an assumption we can make in general, especially when `prepend` is in play:
```ruby
module A; end

class Base
  prepend A
end

class Sub < Base
end

p Sub.ancestors.take_while { |a| Base != a }
# => [Sub, A]
```

Reply:
Ah, I think I know the most performant way to do this, and with no loops at all: Chop the tail of the ancestors list by the size of the ancestor list of the superclass.

If the ancestor list of class `A` is $[a_1, a_2, ..., a_n]$ then, by definition, for class `B < A`, the ancestor list will be $[b_1, b_2, ..., b_m, a_1, a_2, ..., a_n]$. Since what we are trying to do here it to remove all the ancestors that are coming into `B`'s ancestor list from `A`, we can simply chop the $n$ element tail of the ancestor list of `B` by:
```ruby
return ancestors_of(mod)[0...-ancestors_of(superclass)].include?(helper)
```

This is exactly what the relatively recent `is_a?` fast-path optimization in CRuby does, btw: https://www.youtube.com/watch?v=qlq-iQGtCgs

Reply:
Great approach! I applied the same idea with a slight change: `a.take(a.count - i)` is (marginally, but consistently) faster than `a[0...-i]` and `a[...-i]`


<details><summary><code>mini_benchmark.rb</code></summary>

```rb
require 'benchmark/ips'

def ancestors_of(mod) = mod.ancestors
def superclass_of(mod) = mod.superclass

mod = String
superclass = superclass_of(mod)

Benchmark.ips do |x|
  x.config(warmup: 0.1, time: 5)

  x.report("0...-i") do |times|
    i = 0
    while (i += 1) <= times
      ancestors_of(mod)[0...-ancestors_of(superclass).size]
    end
  end
  
  x.report("...-i") do |times|
    i = 0
    while (i += 1) <= times
      ancestors_of(mod)[...-ancestors_of(superclass).size]
    end
  end
  
  x.report("take") do |times|
    i = 0
    while (i += 1) <= times
      a = ancestors_of(mod)
      a.take(a.count - ancestors_of(superclass).size)
    end
  end
  
  x.compare!
end
```

</details>